src/query_ctx.c

```diff
@@ -108,35 +113,101 @@ Graph *QueryCtx_GetGraph(void) {
 
 RedisModuleCtx *QueryCtx_GetRedisModuleCtx(void) {
 	QueryCtx *ctx = _QueryCtx_GetCtx();
-	return ctx->redisctx;
+	return ctx->global_exec_ctx.redis_ctx;
+}
+
+static void _QueryCtx_ThreadSafeContextLock(QueryCtx *ctx) {
+	if(ctx->global_exec_ctx.bc) RedisModule_ThreadSafeContextLock(ctx->global_exec_ctx.redis_ctx);
+}
+
+static void _QueryCtx_ThreadSafeContextUnlock(QueryCtx *ctx) {
+	if(ctx->global_exec_ctx.bc) RedisModule_ThreadSafeContextUnlock(ctx->global_exec_ctx.redis_ctx);
+}
+
+bool QueryCtx_LockForCommit(void) {
+	QueryCtx *ctx = _QueryCtx_GetCtx();
+	// Lock GIL.
+	RedisModuleCtx *redis_ctx = ctx->global_exec_ctx.redis_ctx;
+	GraphContext *gc = ctx->gc;
+	_QueryCtx_ThreadSafeContextLock(ctx);
+	// Open key and verify.
+	RedisModuleString *graphID = RedisModule_CreateString(redis_ctx, gc->graph_name,
+														  strlen(gc->graph_name));
+	RedisModuleKey *key = RedisModule_OpenKey(redis_ctx, graphID, REDISMODULE_WRITE);
+	if(RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY) {
+		asprintf(&ctx->internal_exec_ctx.error, "Encountered an empty key when opened key %s",
+				 ctx->gc->graph_name);
+
+		goto clean_up;
+	}
+	if(RedisModule_ModuleTypeGetType(key) != GraphContextRedisModuleType) {
+		asprintf(&ctx->internal_exec_ctx.error, "Encountered a non-graph value type when opened key %s",
+				 ctx->gc->graph_name);
+		goto clean_up;
+
+	}
+	if(gc != RedisModule_ModuleTypeGetValue(key)) {
+		asprintf(&ctx->internal_exec_ctx.error, "Encountered different graph value when opened key %s",
+				 ctx->gc->graph_name);
+		goto clean_up;
+	}
+	RedisModule_FreeString(redis_ctx, graphID);
+	ctx->internal_exec_ctx.key = key;
+	// Acquire graph write lock.
+	Graph_AcquireWriteLock(gc->g);
+	return true;
+
+clean_up:
+	// Unlock GIL.
+	_QueryCtx_ThreadSafeContextUnlock(ctx);
+	// If there is a break point for runtime exception, raise it, otherwise return false.
+	QueryCtx_RaiseRuntimeException();
+	return false;
+
+}
+
+void QueryCtx_NotifyCommitAndUnlock(void) {
```

Comment:
`QueryCtx_NotifyCommitAndUnlock` is `QueryCtx_LockForCommit` counterpart please name accordingly 
`QueryCtx_UnLockCommit` ???

Reply:
the notify was ment to indicate about the `replicate`. changed

Reply:
Yeah I've got it.